Sources/TSCBasic/FileSystem.swift

```diff
@@ -760,6 +777,21 @@ public class InMemoryFileSystem: FileSystem {
 
         contents.entries[sourcePath.basename] = nil
     }
+
+    public func withLock<T>(on path: AbsolutePath, type: LockType = .exclusive, _ body: () throws -> T) throws -> T {
+        var fileQueue: DispatchQueue
+
+        lockFilesLock.lock()
+        if let queue = lockFiles[path] {
+            fileQueue = queue
+        } else {
+            fileQueue = DispatchQueue(label: "foo", attributes: .concurrent)
+            lockFiles[path] = fileQueue
+        }
+        lockFilesLock.unlock()
```

Comment:
use lock.withLock?

Reply:
If I would use `lock.withLock` I cannot leave `var fileQueue: DispatchQueue` unassigned. Making it optional would mean that I somehow have to handle the case where `fileQueue` is nil.

Reply:
I guess it could be defined as `var fileQueue: DispatchQueue!`, or do we not use this pattern in this repo?

Reply:
We generally try to avoid force unwrap, but there are exceptions if it makes the rest of the code safer and if there's really no way for the force unwrap to fail.

But in this case I think `withLock` is declared as `public func withLock<T>(_ body: () throws -> T) rethrows -> T`, so you should be able to return the `fileQueue` from the `withLock` closure.

Reply:
> But in this case I think withLock on a regular Lock is also declared as public func withLock<T>(_ body: () throws -> T) rethrows -> T, so could you return the fileQueue from the withLock closure?

that's a great idea